core/solver/multigrid.cpp

```diff
@@ -159,346 +160,342 @@ void handle_list(
 }
 
 
+template <typename Vec>
+void clear_and_reserve(Vec& vec, size_type size)
+{
+    vec.clear();
+    vec.reserve(size);
+}
+
+
 }  // namespace
 
 
-struct MultigridState {
-    MultigridState(const LinOp* system_matrix_in, const Multigrid* multigrid_in,
-                   const size_type nrhs_in)
-        : system_matrix(system_matrix_in),
-          multigrid(multigrid_in),
-          nrhs(nrhs_in)
-    {
-        auto current_nrows = system_matrix->get_size()[0];
-        auto mg_level_list = multigrid->get_mg_level_list();
-        auto list_size = mg_level_list.size();
-        auto cycle = multigrid->get_cycle();
-        r_list.reserve(list_size);
-        g_list.reserve(list_size);
-        e_list.reserve(list_size);
-        one_list.reserve(list_size);
-        neg_one_list.reserve(list_size);
-        if (cycle == multigrid::cycle::kfcg ||
-            cycle == multigrid::cycle::kgcr) {
-            kcycle_state.reserve_space(this);
-        }
-        // Allocate memory first such that repeating allocation in each iter.
-        for (int i = 0; i < mg_level_list.size(); i++) {
-            auto next_nrows =
-                mg_level_list.at(i)->get_coarse_op()->get_size()[0];
-            auto mg_level = mg_level_list.at(i);
-
-            run<gko::multigrid::EnableMultigridLevel, float, double,
-                std::complex<float>, std::complex<double>>(
-                mg_level,
-                [this](auto mg_level, auto i, auto cycle, auto current_nrows,
-                       auto next_nrows) {
-                    using value_type = typename std::decay_t<
-                        gko::detail::pointee<decltype(mg_level)>>::value_type;
-                    this->allocate_memory<value_type>(i, cycle, current_nrows,
-                                                      next_nrows);
-                },
-                i, cycle, current_nrows, next_nrows);
-
-            current_nrows = next_nrows;
-        }
-    }
+namespace multigrid {
+
+
+/**
+ * The enum class is to combine the cycle infomation  It's legal to use a binary
+ * or(|) operation to combine several properties.
+ */
+enum class cycle_info {
+    /**
+     * indicate input is zero
+     */
+    x_is_zero = 1,
+
+    /**
+     * current process is the first one of the cycle
+     */
+    first_of_cycle = 2,
+
+    /**
+     * current procees is the end one of the cycle
+     */
+    end_of_cycle = 4
+};
+
+
+GKO_ATTRIBUTES GKO_INLINE cycle_info operator|(cycle_info a, cycle_info b)
+{
+    return static_cast<cycle_info>(static_cast<int>(a) | static_cast<int>(b));
+}
+
+
+GKO_ATTRIBUTES GKO_INLINE bool operator&(cycle_info a, cycle_info b)
```

Comment:
I'm not too happy about this operator, it's a bit counterintuitive to return bool from a bitwise operator.

Reply:
I see. I can change it to `check_property(info, info)`?

Reply:
Yes, a function is definitely more appropriate. I would suggest `has_property`, since that asks more of a question, compared to `check_property`, which describes an action to be taken, so the return value is not as clear.